core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala

```diff
@@ -149,7 +158,11 @@ private[spark] class LiveListenerBus(conf: SparkConf) {
     }
 
     this.sparkContext = sc
-    queues.asScala.foreach(_.start(sc))
+    queues.asScala.foreach { q =>
+      q.start(sc)
+      queuedEvents.foreach(q.post)
```

Comment:
```
q.start(sc)
queuedEvents.foreach(q.post)
```
Ummm... In my opinion, exchange these two lines sequence would be better for following the original logic of events buffered before a queue calls start(). So, queuedEvents post to queues first before queues start would be unified logically.

Reply:
That really does not make any difference in behavior.

Reply:
Yes，that‘s true, semantic only.

Reply:
What if stop() called before all queuedEvents post to AsyncEventQueue?
```
/**
   * Stop the listener bus. It will wait until the queued events have been processed, but new
   * events will be dropped.
   */
```
**(the "queued  events" mentioned in description above is not equal to "queuedEvents" here.)**

As queuedEvents "post" before  listeners install, so, can they be treated  as new events?

Reply:
Ok，**synchronized** can avoid this problem.

Reply:
both `start()` and `stop()` are synchronized so they can't be interleaved.

Reply:
LiveListener's stop() is not synchronized completely. And LiveListener's `stopped` variable could be set "true" in stop() (before the synchronize body) while start() is calling. Do I missing something?

Reply:
sure, but the interleaving you were worried about specifically above isn't possible because of the synchronized blocks.

Yes, you could get one line into `start()`, switch to `stop()` and get a few lines in ... but then `stop()` gets blocked on `start()` anyway.  Sure, that means that while `start()` is running, we also have `stopped == true` ... but so what?  More `post()` calls will be no-ops.  `start()` will start the queues and post all buffered events, then release the lock and `stop()` will stop all the queues.

Reply:
Okay, perfect explanation. Thanks.